proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,582 @@
+`/MyTopic/_intra` topic. Also, the `Subscription` in the second process will be listening to that topic. This
+results in that the meta-message will be delivered to both `Subscription`s. The one in the second topic will
+still try to use the meta-message to retrieve something from the `IntraProcessManager`. In this simple
+scenario it will not be able to find anything and this will only result in a warning message. However, if some
+`Publisher` with intra-process communication enabled happens to be present also in that process, then the
+`Subscription` will end up extracting something from the ring buffer. This is due to the fact that the id of
+the publisher contained in the meta-message is just an integer value guaranteed to be unique only within the
+process that generated it.
+
+This situation has two possible outcomes: if the `Publisher` from which the message is extracted has the same
+message type as the `Subscription`, then the problem will not be noticed, even if the `Subscription` is taking
+a message that can potentially be old or destinated to a different consumer. On the other hand, if the
+`Publisher` from which the message is extracted has a different message type from the one of the
+`Subscription`, the `static_cast` operations performed by the `IntraProcessManager` will cause an access
+violation error.
+
+
+## Proposed implementation
+
+### Overview
+
+The new proposal for intra-process communication addresses all the issues previously mentioned. It has been
+designed with performance in mind, so it avoids any communication through the middleware between nodes in the
+same process. Moreover, it supports all the ROS2 Quality of Service configurations.
+
+Consider a simple scenario, consisting of `Publisher`s and `Subscription`s all in the same process and with
+the durability QoS set to `volatile`. The proposed implementation creates one buffer per `Subscription`. When
+a message is published to a topic, its `Publisher` pushes the message into the buffer of each of the
+`Subscription`s related to that topic and raises a notification, waking up the executor. The executor can then pop
+the message from the buffer and trigger the callback of the `Subscription`.
+
+![Proposed IPC Block Diagram](new_ipc.png)
+
+There are three possible data-types that can be stored in the buffer:
+
+- `MessageT`
+- `shared_ptr<const MessageT>`
+- `unique_ptr<MessageT>`
+
+The choice of the buffer data-type is controlled through an additional field in the `SubscriptionOptions`. The
+default value for this option is `CallbackDefault`, which corresponds to selecting the type between
+`shared_ptr<constMessageT>` and `unique_ptr<MessageT>` that better fits with its callback type. This is
+deduced looking at the output of `AnySubscriptionCallback::use_take_shared_method()`.
+
+If the history QoS is set to `keep all`, the buffers are dynamically allocated. On the other hand, if the
```

Comment:
Mmm could you clarify the difference between `sample` and `instance`? 

See my last comment https://github.com/irobot-ros/ros2-performance/pull/1#issuecomment-501331995, I decoupled all the different components and in this way it will be much simpler to support different options for the length of the queue

Reply:
Here's what the documentation for Connext [1] says :

> Keep all the samples.
>
> On the publishing side, RTI Connext will attempt to keep all samples (representing each value written) of each instance of data (identified by its key) managed by the DDSDataWriter until they can be delivered to all subscribers.
>
> On the subscribing side, RTI Connext will attempt to keep all samples of each instance of data (identified by its key) managed by the DDSDataReader. These samples are kept until the application takes them from RTI Connext via the take() operation. 

@wjwwood seems to be suggesting that there's another limit above which the size of the history does not grow. That's reasonable I think.

In this intra-process proposal, the messages are delivered directly to the subscriber's buffer so publishing side does not need to do anything different. On the sub side, we would have to allow dynamic resizing of the buffer (i.e. standard queue instead of ring buffer) if this QoS is enabled.

Do you agree @wjwwood?

[1] https://community.rti.com/static/documentation/connext-dds/5.3.0/doc/api/connext_dds/api_cpp/group__DDSHistoryQosModule.html#gga96c54dadb2beb87b23776e766e282185a9e526d876dae232aa083125c72135f5b